File: core/interaction/browser_interaction_engine.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

DEFAULT_TIMEOUT_MS = 30000
MAX_ACTIONS = 1000


def record_interaction(
    action: str,
    selector: str,
    metadata: Optional[Dict[str, Any]] = None,
    step: int = 0,
) -> Dict[str, Any]:
    return {
        "id": f"interaction_{step}",
        "timestamp": step,
        "action": str(action).strip(),
        "selector": str(selector).strip(),
        "metadata": dict(metadata or {}),
        "bounded": True,
    }


def build_interaction_plan(
    actions: List[Dict[str, Any]],
) -> Dict[str, Any]:
    normalized: List[Dict[str, Any]] = []

    for index, action in enumerate(actions[:MAX_ACTIONS]):
        normalized.append(
            record_interaction(
                action=str(action.get("type", action.get("action", ""))),
                selector=str(action.get("selector", "")),
                metadata={
                    "value": action.get("value"),
                },
                step=index,
            )
        )

    return {
        "interaction_plan": normalized,
        "bounded": True,
    }


def _bounded_timeout(timeout_ms: Optional[int]) -> int:
    timeout = int(timeout_ms or DEFAULT_TIMEOUT_MS)
    return min(max(timeout, 1), DEFAULT_TIMEOUT_MS)


def click_element(
    page: Any,
    selector: str,
    timeout_ms: Optional[int] = None,
) -> Dict[str, Any]:
    timeout = _bounded_timeout(timeout_ms)

    if page is not None and hasattr(page, "click"):
        page.click(selector, timeout=timeout)

    return {
        "action": "click",
        "selector": selector,
        "timeout_ms": timeout,
        "bounded": True,
    }


def fill_input(
    page: Any,
    selector: str,
    value: str,
    timeout_ms: Optional[int] = None,
) -> Dict[str, Any]:
    timeout = _bounded_timeout(timeout_ms)

    if page is not None and hasattr(page, "fill"):
        page.fill(selector, str(value)[:5000], timeout=timeout)

    return {
        "action": "fill",
        "selector": selector,
        "value": str(value)[:5000],
        "timeout_ms": timeout,
        "bounded": True,
    }


def select_option(
    page: Any,
    selector: str,
    value: str,
    timeout_ms: Optional[int] = None,
) -> Dict[str, Any]:
    timeout = _bounded_timeout(timeout_ms)

    if page is not None and hasattr(page, "select_option"):
        page.select_option(selector, str(value)[:5000], timeout=timeout)

    return {
        "action": "select",
        "selector": selector,
        "value": str(value)[:5000],
        "timeout_ms": timeout,
        "bounded": True,
    }


def hover_element(
    page: Any,
    selector: str,
    timeout_ms: Optional[int] = None,
) -> Dict[str, Any]:
    timeout = _bounded_timeout(timeout_ms)

    if page is not None and hasattr(page, "hover"):
        page.hover(selector, timeout=timeout)

    return {
        "action": "hover",
        "selector": selector,
        "timeout_ms": timeout,
        "bounded": True,
    }


def wait_for_selector(
    page: Any,
    selector: str,
    timeout_ms: Optional[int] = None,
) -> Dict[str, Any]:
    timeout = _bounded_timeout(timeout_ms)

    if page is not None and hasattr(page, "wait_for_selector"):
        page.wait_for_selector(selector, timeout=timeout)

    return {
        "action": "wait",
        "selector": selector,
        "timeout_ms": timeout,
        "bounded": True,
    }

File: src/interaction/browserInteractionEngine.ts
/**
 * Converted from Python: core/interaction/browser_interaction_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let DEFAULT_TIMEOUT_MS: any = 30000;
export let MAX_ACTIONS: any = 1000;
export function recordInteraction(action: any, selector: any, metadata: any = null, step: any = 0): any {
  return {"id": `interaction_${py.toStr(step)}`, "timestamp": step, "action": py.strip(py.toStr(action)), "selector": py.strip(py.toStr(selector)), "metadata": py.pyDict(py.or2(metadata, () => ({}))), "bounded": true};
}
export function buildInteractionPlan(actions: any): any {
  var normalized: any[] = [];
  var index: any;
  var action: any;
  for ([index, action] of py.enumerate(py.slice(actions, null, MAX_ACTIONS))) {
    py.listAppend(normalized, recordInteraction(py.toStr(py.get(action, "type", py.get(action, "action", ""))), py.toStr(py.get(action, "selector", "")), {"value": py.get(action, "value")}, index));
  }
  return {"interaction_plan": normalized, "bounded": true};
}
export function _boundedTimeout(timeout_ms: any): any {
  var timeout: any = py.toInt(py.or2(timeout_ms, () => (DEFAULT_TIMEOUT_MS)));
  return py.min([py.max([timeout, 1]), DEFAULT_TIMEOUT_MS]);
}
export function clickElement(page: any, selector: any, timeout_ms: any = null): any {
  var timeout: any = _boundedTimeout(timeout_ms);
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("click") in (page as object) || typeof (page as Record<string, unknown>)[String("click")] === "function")))) {
    page.click(selector, timeout);
  }
  return {"action": "click", "selector": selector, "timeout_ms": timeout, "bounded": true};
}
export function fillInput(page: any, selector: any, value: any, timeout_ms: any = null): any {
  var timeout: any = _boundedTimeout(timeout_ms);
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("fill") in (page as object) || typeof (page as Record<string, unknown>)[String("fill")] === "function")))) {
    page.fill(selector, py.slice(py.toStr(value), null, 5000), timeout);
  }
  return {"action": "fill", "selector": selector, "value": py.slice(py.toStr(value), null, 5000), "timeout_ms": timeout, "bounded": true};
}
export function selectOption(page: any, selector: any, value: any, timeout_ms: any = null): any {
  var timeout: any = _boundedTimeout(timeout_ms);
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("select_option") in (page as object) || typeof (page as Record<string, unknown>)[String("select_option")] === "function")))) {
    page.select_option(selector, py.slice(py.toStr(value), null, 5000), timeout);
  }
  return {"action": "select", "selector": selector, "value": py.slice(py.toStr(value), null, 5000), "timeout_ms": timeout, "bounded": true};
}
export function hoverElement(page: any, selector: any, timeout_ms: any = null): any {
  var timeout: any = _boundedTimeout(timeout_ms);
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("hover") in (page as object) || typeof (page as Record<string, unknown>)[String("hover")] === "function")))) {
    page.hover(selector, timeout);
  }
  return {"action": "hover", "selector": selector, "timeout_ms": timeout, "bounded": true};
}
export function waitForSelector(page: any, selector: any, timeout_ms: any = null): any {
  var timeout: any = _boundedTimeout(timeout_ms);
  if (((page !== null && page !== undefined) && (page !== null && page !== undefined && typeof page === "object" && (String("wait_for_selector") in (page as object) || typeof (page as Record<string, unknown>)[String("wait_for_selector")] === "function")))) {
    page.wait_for_selector(selector, timeout);
  }
  return {"action": "wait", "selector": selector, "timeout_ms": timeout, "bounded": true};
}
